### PR TITLE
validate input shapes

### PR DESCRIPTION
**Pull request by jbkyang-nvi**
Fix: https://github.com/triton-inference-server/server/issues/7171
Validate input shapes at core level
Testing: https://github.com/triton-inference-server/server/pull/7195

### CODE REVIEW

File: src/infer_request.cc [hunk 1]
```diff
@@ -1134,6 +1135,24 @@ InferenceRequest::Normalize()
...
       }
     }
+    // Matching incoming request's shape and byte size to make sure the
+    // payload contains correct number of elements
+    {
+      const size_t& byte_size = input.Data()->TotalByteSize();
+      const auto& data_type = input.DType();
+      const auto& input_dims = *shape;
+      const int64_t expected_byte_size =
+          triton::common::GetByteSize(data_type, input_dims);
```
Comment (Tabrizian) [inline src/infer_request.cc:1145]: Note that this API for `TYPE_STRING` returns 0 and we might some custom handling for that.

File: src/infer_request.cc [hunk 2]
```diff
@@ -1169,6 +1170,79 @@ InferenceRequest::Normalize()
...
+    {
+      const size_t& byte_size = input.Data()->TotalByteSize();
+      const auto& data_type = input.DType();
+      const auto& input_dims = input.ShapeWithBatchDim();
+      int64_t expected_byte_size = INT_MAX;
+      // Because Triton expects STRING type to be in special format
+      // (prepend 4 bytes to specify string length), so need to add all the
+      // first 4 bytes for each element to find expected byte size
+      if (data_type == inference::DataType::TYPE_STRING) {
+        int64_t element_count = triton::common::GetElementCount(input_dims);
```
Comment (GuanLuo) [inline src/infer_request.cc:1186]: Wrap in helper function (`ValidateBytesInput`) for readability

File: src/infer_request.cc [hunk 3]
```diff
@@ -1169,6 +1170,79 @@ InferenceRequest::Normalize()
...
+        int64_t element_idx = 0;
+        expected_byte_size = 0;
+        for (size_t i = 0; i < input.Data()->BufferCount(); ++i) {
+          BufferAttributes* buffer_attributes;
+          const char* content = input.Data()->BufferAt(i, &buffer_attributes);
+          size_t content_byte_size = input.Data()->TotalByteSize();
+
+          while (content_byte_size >= sizeof(uint32_t)) {
```
Comment (tanmayv25) [inline src/infer_request.cc:1194]: I don't think this logic scales well for the case where `BufferCount() > 1`. The `content_byte_size` would return the bytesize for the complete input tensor including all the buffers.

Comment (yinggeh) [reply src/infer_request.cc:1196]: Fixed

File: src/infer_request.cc (hunk 2 above)
Comment (yinggeh) [reply src/infer_request.cc]: Fixed

File: src/infer_request.cc (hunk 1 above)
Comment (yinggeh) [reply src/infer_request.cc]: Fixed

**Review by GuanLuo — APPROVED** [review]

**Comment by yinggeh** [comment]
Do not merge as seeing failures in [L0_cuda_shared_memory](https://gitlab-master.nvidia.com/dl/dgx/tritonserver/-/jobs/93001511) and [L0_shared_memory](https://gitlab-master.nvidia.com/dl/dgx/tritonserver/-/jobs/93020179)

**Comment by rmccorm4** [comment]
Please don't merge a PR before the corresponding tests or related PRs are also passing and ready to merge. It looks like the testing is still a WIP: https://github.com/triton-inference-server/server/pull/7195.

**Pull request merged**